utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -105,9 +95,18 @@ class BestEffortJsonSchemaEncoder(validationMode: ValidationMode) {
   private def error(str: String): Invalid[NonEmptyList[String]] = Invalid(NonEmptyList.of(str))
 
   private def error(runtimeType: String, schema: String, fieldName: Option[String]): Invalid[NonEmptyList[String]] =
-    Invalid(NonEmptyList.of(s"Not expected type: ${runtimeType} for field${fieldName.map(f => s": '$f'").getOrElse("")} with schema: $schema"))
+    Invalid(NonEmptyList.of(s"Not expected type: $runtimeType for field${fieldName.map(f => s": '$f'").getOrElse("")} with schema: $schema."))
+
+  private[encode] def encodeWithJsonValidation(value: Any, schema: Schema, fieldName: Option[String] = None): EncodeOutput =
+    encode(value, schema, fieldName).andThen { result =>
+      val json = JsonSchemaUtils.circeToJson(result)
+      schema
+        .validateData(json)
+        .leftMap(errorMsg => NonEmptyList.of(errorMsg))
+        .map(_ => result) //we return base json (e.g. without felt defaults - there is no need to put these data as output)
```

Comment:
in [deserialization](https://github.com/TouK/nussknacker/pull/3742/files#diff-e512d01d8979f3d4c273889b9675a818ade23b7488403e051c1961340d35fa8bR36)   we take validated object, and here not, why?

Reply:
Because when you do deserialization you want to have all data in HashMap -> e.g. after source you can use field with default and this field should be felt. At sink we want to have optimized object (there is no need to fill by defaults etc. because this object is still valid)